### PR TITLE
[CONJ-805] >1Byte UTF8 character issue if maxFieldSize is set

### PR DESCRIPTION
**Pull request by MarkusLutum**
Detected at one of my clients.
They use polish special characters like ść
The length in TextRowProtocol is set to 2 for each of those characters so 4 in our example.
With this the current code 
`return new String(buf, pos, Math.min(maxFieldSize * 3, length), StandardCharsets.UTF_8)
                    .substring(0, Math.min(maxFieldSize, length));`
results into an ArrayOutOfBoundsError at the  substring call.

I added a test case and a quick fix for it.
Travis fails on environment issues. Nothing to do with my change.

### CODE REVIEW

**Comment by MarkusLutum** [comment]
Files a bug for this as well:
https://jira.mariadb.org/browse/CONJ-805

File: src/main/java/org/mariadb/jdbc/internal/com/read/resultset/rowprotocol/TextRowProtocol.java [hunk 1]
```diff
@@ -234,8 +234,9 @@ public String getInternalString(ColumnDefinition columnInfo, Calendar cal, TimeZ
     }
 
     if (maxFieldSize > 0) {
-      return new String(buf, pos, Math.min(maxFieldSize * 3, length), StandardCharsets.UTF_8)
-          .substring(0, Math.min(maxFieldSize, length));
+    	String value = new String(buf, pos, Math.min(maxFieldSize * 3, length), StandardCharsets.UTF_8);
+    	int maxLength = Math.min(maxFieldSize, length);
+    	return  value.substring(0, Math.min(maxLength, value.length()));
```
Comment (rusher) [inline src/main/java/org/mariadb/jdbc/internal/com/read/resultset/rowprotocol/TextRowProtocol.java:239]:
```
int maxLength = Math.min(maxFieldSize, length);
return value.substring(0, Math.min(maxLength, value.length()));
```
can be simplified to 

```
if (maxFieldSize < value.length()) {
   return value.substring(0, maxFieldSize);
}
return value;
```
value.length will always be smaller or equal than length.

Comment (MarkusLutum) [reply src/main/java/org/mariadb/jdbc/internal/com/read/resultset/rowprotocol/TextRowProtocol.java]: Correct. Changed it to your suggestion.

**Comment by harawata** [comment]
Both the current implementation and the proposed fix count (or try to count) the length of the string, however, the [doc](https://docs.oracle.com/en/java/javase/14/docs/api/java.sql/java/sql/Statement.html#setMaxFieldSize(int)) says...

> Sets the limit for the maximum number of bytes that can be returned for character and binary column values in a ResultSet object produced by this Statement object.

So, it should look something like this, shouldn't it?

```java
if (maxFieldSize > 0) {
  return new String(buf, pos, maxFieldSize, StandardCharsets.UTF_8);
}
```

**Comment by rusher** [comment]
good point @harawata 

In JDBC from [spec ](https://download.oracle.com/otndocs/jcp/jdbc-4_3-mrel3-spec/index.html):

> 8.3.1 Silent Truncation
> The Statement.setMaxFieldSize method allows a maximum size (in bytes) to
> be set. This limit applies only to the BINARY, VARBINARY, LONGVARBINARY, CHAR,
> VARCHAR, LONGVARCHAR, NCHAR, NVARCHAR, and LONGNVARCHAR data types.
> If a limit has been set using setMaxFieldSize and there is an attempt to read data
> that exceeds the limit, any truncation that occurs as a result of exceeding the set limit
> will not be reported.

CHAR and VARCHAR are explicitly listed.

This is then not just a bug correction, but a change of behavior compares to current driver implementation, so has then to be in the next 2.7 version, not in a corrective patch version.

**Comment by MarkusLutum** [comment]
So what to do @rusher. Should I quickly change my pull request to @harawata s suggestion? And you do the rest?

**Pull request merged**